PowerNSX.psm1

```diff
@@ -21191,6 +21191,432 @@ function Remove-NsxSecurityGroupMember {
+
+    $sg3 = New-NsxSecurityGroup -Name "SG-Backup-Clients"
+
+    $sg3.objectid | Add-NsxDymanicMemberSet -operator OR -DynamicCriteriaSpec $criteria5Spec,$criteria6Spec
+    $sg3.objectid | Add-NsxDymanicMemberSet -operator OR -DynamicCriteriaSpec $criteria7Spec,$criteria8Spec
+
+    #>
+
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSAvoidDefaultValueSwitchParameter","")] # Cant remove without breaking backward compatibility
+
+    param (
+        [Parameter (Mandatory=$true, ValueFromPipeline=$true, Position=1) ]
+            # SecurityGroup whose membership is to be modified.
+            [ValidateNotNullOrEmpty()]
+            [object]$SecurityGroup,
+        [Parameter (Mandatory=$true) ]
+            [ValidateSet("OR", "AND", IgnoreCase=$false)]
+            [String]$operator,
+        [Parameter (Mandatory=$true) ]
+            # Dynamic criteria spec/s as generated by New-NsxDynamicCriteriaSpec
+            [ValidateNotNullOrEmpty()]
+            [System.Xml.XmlElement[]]$DynamicCriteriaSpec,
+        [Parameter (Mandatory=$False)]
+            #PowerNSX Connection object
+            [ValidateNotNullOrEmpty()]
+            [PSCustomObject]$Connection=$defaultNSXConnection
+    )
+
+    begin {
+    }
+
+    process {
+        # Get our internal SG object and id.  The internal obejct is used later
+        # to modify and put for bulk update.
+        if ( $SecurityGroup -is [System.Xml.XmlElement] ) {
+            $SecurityGroupId = $securityGroup.objectId
+            $_SecurityGroup = $SecurityGroup.cloneNode($true)
+        }
+        elseif ( ($securityGroup -is [string]) -and ($SecurityGroup -match "securitygroup-\d+")) {
+            $SecurityGroupId = $securityGroup
+            $_SecurityGroup = Get-NsxSecurityGroup -objectId $SecurityGroupId -connection $connection
+        }
+        else {
+            throw "Invalid SecurityGroup specified.  Specify a PowerNSX SecurityGroup object or a valid securitygroup objectid."
+        }
+
+        # First we need to verify if the Security Group object passed in via the
+        # pipeline already has the dynamicMemberDefinition element created. If
+        # not, then create the required XML structure
+        if ( -not (Invoke-XPathQuery -QueryMethod SelectSingleNode -Node $_SecurityGroup -Query 'child::dynamicMemberDefinition') ) {
+            Add-XmlElement -xmlRoot $_SecurityGroup -xmlElementName "dynamicMemberDefinition"
+        }
+
+        # Now lets add the dynamic criteria (DynamicSets)
+        # TODO: Will need to put some input validation on the DynamicCriteriaSpec parameter to make sure what is being specified is actually a Dynamic Criteria Spec
+        if ( $PsBoundParameters.ContainsKey('DynamicCriteriaSpec') ) {
+            [System.Xml.XmlElement]$xmlDynamicMemberDefinition = invoke-xpathquery -QueryMethod SelectSingleNode -node $_SecurityGroup -query "child::dynamicMemberDefinition"
```

Comment:
youve already queried for this - you should query once - if it does exists, you have your object, if it doesnt, then create and your returned object from above (part2)